docs/data-sourcing-and-transformation.md

```diff
@@ -0,0 +1,91 @@
+
+`data` templates can best be understood by looking at a common data sourcing and transformation operation in `bash`:
+
+```bash
+$ find -r . | grep ".pdf" | sed "s/foo/foo.ready/"
+```
+
+Such operations consist of two main parts:
+* A "source" of data: `find -r .`
+* A series of "transformations" which transform the output of the source serially: `| grep ".pdf" | sed "s/foo/foo.ready/"`
+
+This operation, for example, could be useful in sourcing a potential list of files to be processed and filtering and manipulating the list as desired.
+
+In Argo, this operation would be written as:
+
+```yaml
+- name: generate-artifacts
+  data:
+    source:             # Define a source for the data, only a single "source" is permitted
+      artifactPaths:    # A predefined source: Generate a list of all artifact paths in a given repository
+        s3:             # Source from an S3 bucket
+          bucket: test
+          endpoint: minio:9000
+          insecure: true
+          accessKeySecret:
+            name: my-minio-cred
+            key: accesskey
+          secretKeySecret:
+            name: my-minio-cred
+            key: secretkey
+    transformation:     # The source is then passed to be transformed by transformations defined here
+      - expression: "filter(data, {# endsWith \".pdf\"})"
+      - expression: "map(data, {# + \".ready\"})"
+```
+
+## In-Memory Transformations
+
+If possible, Argo will choose to perform the specified transformation in its own memory, saving the need to create a pod in the cluster for said transformation. Because of this, in-memory transformations are near instantaneously completed (much like a [cached/memoized node](memoization.md)).
+
+Currently, Argo will always choose to perform the transformation in memory if it does not use `artifactPaths` to source the data.
+
+However, if you would like to force Argo to offload the transformation to a Pod regardless, you can specify it by using `podPolicy: "Always"`:
+
+```yaml
+- name: generate-artifacts
+  data:
+    podPolicy: "Always"
+    source:
+      ...
+```
+
+If you are the manager of the Argo controller and you would like to force your users to always offload transformations to pods, you can specify so in [workflow restrictions](workflow-restrictions.md). This can be used as a security or prevention measure to protect the controller from large transformations that may explode its memory usage.
+
+## Spec
+
+A `data` template must always contain a `source`. Current available sources:
+
+* `artifactPaths`: generates a list of artifact paths from the artifact repository specified
+* `raw`: reads raw input. This is useful when passing data to the template as an input parameter:
+    
+    ```yaml
+    - name: collect-artifacts
+      inputs:
+        parameters:
+          - name: processed
+      data:
+        source:
+          raw: "{{inputs.parameters.processed}}"
```

Comment:
I'm not sure about `raw`, maybe `value` @jessesuen ?

Reply:
`raw` would be consistent with other parts of the code, but I have heard Jesse express regret at this name. I'll bring it up with him

Reply:
did we remove this?

Reply:
Done